pkg/runtime/pubsub/bulkpub.go

```diff
@@ -0,0 +1,111 @@
+/*
+Copyright 2022 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pubsub
+
+import (
+	"sync"
+
+	contribPubsub "github.com/dapr/components-contrib/pubsub"
+)
+
+const (
+	bulkPublishSeriallyKey string = "bulkPublishSerially"
+)
+
+// defaultBulkPublisher is the default implementation of BulkPublisher.
+// It is used when the component does not implement BulkPublisher.
+type defaultBulkPublisher struct {
+	p contribPubsub.PubSub
+}
+
+// NewDefaultBulkPublisher returns a new defaultBulkPublisher from a PubSub.
+func NewDefaultBulkPublisher(p contribPubsub.PubSub) *defaultBulkPublisher {
+	return &defaultBulkPublisher{
+		p: p,
+	}
+}
+
+// BulkPublish publishes a list of messages to a topic as individual Publish requests.
+// If 'bulkPublishSerially' metadata is set to true, the messages are sent to the broker serially,
+// in the same order. Otherwise they are sent to the broker as parallel Publish requests.
+func (p *defaultBulkPublisher) BulkPublish(req *contribPubsub.BulkPublishRequest) (contribPubsub.BulkPublishResponse, error) {
+	if req.Metadata[bulkPublishSeriallyKey] == "true" {
+		return p.bulkPublishSerial(req)
+	} else {
+		return p.bulkPublishParallel(req)
+	}
+}
+
+// bulkPublishSerial publishes messages in a serial order. This is slower, but ensures
+// that messages are sent to the broker in the same order as specified in the request.
+func (p *defaultBulkPublisher) bulkPublishSerial(req *contribPubsub.BulkPublishRequest) (contribPubsub.BulkPublishResponse, error) {
+	statuses := make([]contribPubsub.BulkPublishResponseEntry, 0, len(req.Entries))
+
+	for _, entry := range req.Entries {
+		statuses = append(statuses, p.bulkPublishSingleEntry(req, entry))
```

Comment:
i think this will still allocate new mem every time ... 
should it be 
```suggestion
		statuses[i] = p.bulkPublishSingleEntry(req, entry)
```

Reply:
The capacity is already set while initializing, this won't cause any further allocations.